feat: Subscription size backend integration

MrMuzyk opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

Integrated Subscription Size parts with backend. Data now comes from API and proper API calls are performed when executed

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->

$ https://github.com/Expensify/App/issues/38631


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

To access this newly created component, paste the following link into the browser

https://dev.new.expensify.com:8082/settings/subscription

or add this effect to InitialSettingsPage.tsx

```
useEffect(() => {
    Navigation.navigate(ROUTES.SETTINGS_SUBSCRIPTION)
}, [])
```

- Try to change subscription size to lower/higher number
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

```
useEffect(() => {
    Navigation.navigate(ROUTES.SETTINGS_SUBSCRIPTION)
}, [])
```

- Try to change subscription size to lower/higher number
- [x] Verify that no errors appear in the JS console
### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos

https://github.com/MrMuzyk/App/assets/23487612/7805c596-236e-48eb-a253-0be9af44bea8

MrMuzyk committed:
feat: subscription size backend integration

MrMuzyk committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/subscription-size-backend-integration

MrMuzyk committed:
feat: draft

MrMuzyk committed:
Merge branch 'feat/subscription-size-action' of github.com:MrMuzyk/App into feat/subscription-size-backend-integration

MrMuzyk committed:
fix: small corrections

MrMuzyk committed:
Merge branch 'feat/subscription-size-action' of github.com:MrMuzyk/App into feat/subscription-size-backend-integration

MrMuzyk committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/subscription-size-backend-integration

mananjadhav commented:
Reviewing this now. @MrMuzyk can you resolve the conflicts?

mananjadhav commented on src/libs/actions/Subscription.ts:
```diff
@@ -8,7 +13,53 @@ function openSubscriptionPage() {
+                key: ONYXKEYS.NVP_PRIVATE_SUBSCRIPTION,
+                value: {
+                    userCount: newSubscriptionSize,
+                    pendingFields: {
```
let's reset the error fields to null as well?

mananjadhav commented:
Overall looks great, just one question on the optimistic error handling.

mananjadhav commented:
@MrMuzyk I can see a couple small issues:

1. After updating the subscription size,  when we redirect to `settings/subscription` the query string `canChangeSize=1` isn't removed from the URL.
2. If I open https://dev.new.expensify.com:8082/settings/subscription/subscription-size?canChangeSize=0, and then manually change value of `canChangeSize`, `0` to `1` the text input is blank. Ideally it shouldn't be blank right?
3. Clicking on `X` (close) for the RBR doesn't do anything.
4. Setting the same value as the existing `userCount`, still shows an error message that we can't downgrade the plan.
5. Going offline to online, shows several API failures with 404 error. 


https://github.com/Expensify/App/assets/3069065/8f25602c-4a36-4db0-9f97-6481fa247335

https://github.com/Expensify/App/assets/3069065/cb048848-9e92-440b-be02-08a0110333a4

https://github.com/Expensify/App/assets/3069065/b0cc9e90-fb2c-4e97-b11d-4876b93eb04a

mananjadhav commented:
Also can you correct the typo:

> Integrated Subscription Size parts with backend. Data now comes from API and proper API calls are performed when exectued

Integrated Subscription Size parts with backend. Data now comes from API and proper API calls are performed when _executed_

MrMuzyk committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/subscription-size-backend-integration

MrMuzyk committed:
fix: cr fixes

MrMuzyk commented:
@mananjadhav 

Rebased and fixed conflicts, added error clear on success and fixed the typo.

1. Is it a problem? I doesn't affect page behaviour in any way. Everything works correctly and if you navigate to some other route that isn't related to subscription it disappears.
2. I wasn't able to reproduce that behaviour. Did you do anything special in order to achieve that?

https://github.com/Expensify/App/assets/23487612/af60abd8-6468-48e5-b6c3-f26c5edb91dd

3. Did you mean clicking X on error message? Added handler for that
4. Asked for guidelines on this as Im not sure how it should behave
5. I couldn't reproduce that as well

https://github.com/Expensify/App/assets/23487612/75858581-3d0f-4478-9a97-846037b37e4e

MrMuzyk committed:
fix: handle same size case

MrMuzyk committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/subscription-size-backend-integration

MrMuzyk commented:
@mananjadhav 

4. We're now displaying special error message in case user tries to proceed without changing current subscription size

MrMuzyk committed:
fix: linter

MrMuzyk committed:
fix: linter

MrMuzyk commented:
There is one issue that became visible after merging with main but is unrelated to my changes. When navigating between the screens content shifts - Im trying to find a cause but the fix might come in next PR

https://github.com/Expensify/App/assets/23487612/ecbbb0c4-9f0e-428f-bd51-c9fc6910b8c9

MrMuzyk committed:
fix: exclude subscription screen from focus trap

MrMuzyk commented:
Found the cause - had to exclude subscription ROOT screen from FocusTrap. Everything should behave correctly now

mananjadhav reviewed (commented):
Overall the code changes look fine to me. I'll get started on the checklist and testing.

mananjadhav commented:
@MrMuzyk @amyevans We're showing all the errors in red as an input error.


<img width="798" alt="Screenshot 2024-06-13 at 9 19 17 PM" src="https://github.com/Expensify/App/assets/3069065/d8042209-d840-4882-b28a-bd7c86882fc7">

<img width="805" alt="Screenshot 2024-06-13 at 9 19 11 PM" src="https://github.com/Expensify/App/assets/3069065/38a4126c-c490-4f41-81e9-70966009430e">

<img width="1212" alt="Screenshot 2024-06-13 at 9 19 01 PM" src="https://github.com/Expensify/App/assets/3069065/62998fce-0d51-4ceb-82d5-0b4ce332c7ac">


But when we enter the subscription size count lesser than the current value we show the error in the RHS as a separate screen. Does make sense to add this also as red error message?

<img width="553" alt="Screenshot 2024-06-13 at 9 19 31 PM" src="https://github.com/Expensify/App/assets/3069065/7b794e72-cd4a-4574-a772-c62b4b73bdf1">

mananjadhav commented:
Also let's add a FE validation to only allow integers? Although API doesn't accept decimals (2.05, etc.) but FE allows it.

<img width="380" alt="Screenshot 2024-06-13 at 9 32 30 PM" src="https://github.com/Expensify/App/assets/3069065/d36dbe4b-ee53-4156-b98b-c97abf87da63">

mananjadhav commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/3069065/fa035c14-707b-4c6f-a0de-666bc7fa110a

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/3069065/4cff907b-510b-4ad1-8256-ba6fd5d985e2

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/3069065/0ef693df-b3b8-4efc-8a88-adc7c4f8ea8b

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/3069065/2944d06b-b8d6-4005-b678-69c43adafaed

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/3069065/cffdc05f-874a-4edc-ac89-6475cd355633

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/3069065/b106f0f4-903a-4c4e-b158-ed6111915c1e

</details>

mananjadhav commented:
Just final edge cases. @amyevans Can you review in parallel?

amyevans commented:
> @MrMuzyk @amyevans We're showing all the errors in red as an input error.
> 
> <img alt="Screenshot 2024-06-13 at 9 19 17 PM" width="200" src="https://private-user-images.githubusercontent.com/3069065/339450702-d8042209-d840-4882-b28a-bd7c86882fc7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyOTczMjgsIm5iZiI6MTcxODI5NzAyOCwicGF0aCI6Ii8zMDY5MDY1LzMzOTQ1MDcwMi1kODA0MjIwOS1kODQwLTQ4ODItYjI4YS1iZDdjODY4ODJmYzcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTNUMTY0MzQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzM3YjdjMzBjNmYxM2FlNDI4OTZkNjBiYzI4M2ViODFkNTk1M2RiZTdjNDFkZjk2ODRmNDNkMzg4ZWU0OTUwYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.scauHCg-S7bpHtshl22C_4KaRi8sBKVrPkvkqOQdCB0"> <img alt="Screenshot 2024-06-13 at 9 19 11 PM" width="200" src="https://private-user-images.githubusercontent.com/3069065/339450791-38a4126c-c490-4f41-81e9-70966009430e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyOTczMjgsIm5iZiI6MTcxODI5NzAyOCwicGF0aCI6Ii8zMDY5MDY1LzMzOTQ1MDc5MS0zOGE0MTI2Yy1jNDkwLTRmNDEtODFlOS03MDk2NjAwOTQzMGUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTNUMTY0MzQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDM1MmNmYmE1MjA2ZDJmNjUzZWY5OTEwYzJjZjdlOGY1NmVkY2VjZDg5YzVmOWM5ZTZmYWRkMzQ1MjMyNDUzNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.RqmtRUZ7HtpWgHGranELLNN20tTt0QDxps3acveqBMQ"> <img alt="Screenshot 2024-06-13 at 9 19 01 PM" width="200" src="https://private-user-images.githubusercontent.com/3069065/339450803-62998fce-0d51-4ceb-82d5-0b4ce332c7ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyOTczMjgsIm5iZiI6MTcxODI5NzAyOCwicGF0aCI6Ii8zMDY5MDY1LzMzOTQ1MDgwMy02Mjk5OGZjZS0wZDUxLTRjZWItODJkNS0wYjRjZTMzMmM3YWMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTNUMTY0MzQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDlhMDhmYTcxYzQzZTI5ZTliMGEzYjkwMWJlMzc0MGYwMjgwYmE2OGFjOWNkOWJhM2Y5OGM3NTRiNDJiZDA4YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.d741g_H4RgGlxl0wkIpsFMHtET6Fv48NAivQTTXO5TI">
> But when we enter the subscription size count lesser than the current value we show the error in the RHS as a separate screen. Does make sense to add this also as red error message?
> 
> <img alt="Screenshot 2024-06-13 at 9 19 31 PM" width="553" src="https://private-user-images.githubusercontent.com/3069065/339451642-7b794e72-cd4a-4574-a772-c62b4b73bdf1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyOTczMjgsIm5iZiI6MTcxODI5NzAyOCwicGF0aCI6Ii8zMDY5MDY1LzMzOTQ1MTY0Mi03Yjc5NGU3Mi1jZDRhLTQ1NzQtYTc3Mi1jNjJiNGI3M2JkZjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTNUMTY0MzQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGE3MTA1M2IyYTc4OTUyZWZjYTc0YjBhNjYxZDgxOWFhYmRkYzA2MjRiMGQxODI0YTE5MzllNzNlNGY3MzhiMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.rLgq3inw4ThyZVV6IWL7DAqg0xdUGX_eBqBocAtELqg">

Hmm let's get @dannymcclain's input on this one

dannymcclain commented:
No I don't think we want to use red on that second screen. We use red for validation errors and violations to call your attention to the message, but the text on that screen is isolated and just explaining _why_ you can't do that action. It's informational and there's nothing else competing for it's attention, so we should just leave it as normal text.

We do need to add a period at the end of the first sentence though 🙃 
![CleanShot 2024-06-13 at 12 18 42@2x](https://github.com/Expensify/App/assets/7516624/de330340-a80f-400c-95fd-25862f4150a7)

amyevans commented:
Well I think the suggestion was to totally remove the second screen and move all this text to be an inline error. But that makes sense that we'd treat it differently than a true "error" and leave it as currently implemented 👍 

@MrMuzyk can you add the period to the end of the sentence please?

amyevans commented on src/libs/actions/Subscription.ts:
```diff
@@ -61,4 +63,64 @@ function updateSubscriptionType(type: SubscriptionType) {
+                key: ONYXKEYS.NVP_PRIVATE_SUBSCRIPTION,
+                value: {
+                    userCount: currentSubscriptionSize,
+                    errorFields: {
+                        userCount: ErrorUtils.getMicroSecondOnyxError('common.genericErrorMessage'),
+                    },
```
```suggestion
```

Let's not add the generic error message in the frontend data. Forcing an error from my backend, this is the result:


https://github.com/Expensify/App/assets/7277067/5b35e667-1d0a-4938-84d4-f58ff1bc3073

Because the microtimes are different, both errors display which is more confusing than helpful.

If we remove the generic error we achieve the desired result:

https://github.com/Expensify/App/assets/7277067/f3ae23da-239f-4ab0-9627-0d08caa36346

Also **NOTE**: I need to make a small backend update to send the errors in `['errorFields', 'userCount']` (currently it is sending it in `errors` key, but I agree with your implementation that we need to send it in `['errorFields', 'userCount']`, otherwise we can't distinguish where exactly on the Subscription screen to display the error). cc @blimpich, I'm assigning you for review on that PR

amyevans replied:
The `errorFields` backend change is now on staging by the way

MrMuzyk replied:
Sure thing, changing it

amyevans commented on src/pages/settings/Subscription/SubscriptionSize/substeps/Confirmation.tsx:
```diff
@@ -6,31 +7,37 @@ import FixedFooter from '@components/FixedFooter';
-    const SUBSCRIPTION_UNTIL = subscriptionRenewalDate;
+    const isTryingToIncreaseSubscriptionSize = (subscriptionSizeFormDraft ? Number(subscriptionSizeFormDraft[INPUT_IDS.SUBSCRIPTION_SIZE]) : 0) > (privateSubscription?.userCount ?? 0);
+    const canChangeSubscriptionSize = (account?.canDowngrade ?? false) || (isTryingToIncreaseSubscriptionSize && isEditing);
+    const newSubscriptionEndDate = privateSubscription?.endDate ? format(new Date(privateSubscription?.endDate), CONST.DATE.MONTH_DAY_YEAR_ABBR_FORMAT) : '';
```
Also: the new end date has no relation to the previous end date. The new end date should be based on the current date. So if the user makes this change anytime in June of 2024, their new end date is June 1st 2025 (even if their previous end date was say, Feb 1 2025). If they make the change 7/4/24, their new end date is 7/1/25. And so on 😄

MrMuzyk replied:
This is incorrectly named, should be `formattedSubscriptionEndDate`. Im fixing it and will also take the timezone note into consideration

MrMuzyk replied:
I've created a new util to correctly format subscription date and used it in proper places

amyevans replied:
Perfect, thanks

amyevans reviewed (changes requested):
A few things to address from my perspective as well. Thanks!

MrMuzyk commented on src/pages/settings/Subscription/SubscriptionDetails/index.tsx:
```diff
@@ -42,25 +42,38 @@ function SubscriptionDetails() {
     const [account] = useOnyx(ONYXKEYS.ACCOUNT);
 
     const onOptionSelected = (option: SubscriptionType) => {
+        if (privateSubscription?.type === CONST.SUBSCRIPTION.TYPE.ANNUAL && option === CONST.SUBSCRIPTION.TYPE.PAYPERUSE) {
+            Navigation.navigate(ROUTES.SETTINGS_SUBSCRIPTION_SIZE.getRoute(0));
+            return;
+        }
```
I'm not sure how is it possible that this variation got displayed. If annual option is currently selected and user clicks on pay-per-use he should always get the `youCantDowngrade` screen.

We're navigating him there with proper route param

`Navigation.navigate(ROUTES.SETTINGS_SUBSCRIPTION_SIZE.getRoute(0));`

https://github.com/Expensify/App/assets/23487612/2867c6c2-a07e-40bb-b7c8-15cc070c3c65


I will also add the mentioned check of canDowngrade flag in that condition

MrMuzyk replied:
Added the mentioned check of canDowngrade flag in that condition

amyevans replied:
> Can you reliably reproduce this case? As you can see on the video above it is working fine for me

I can no longer reproduce at all 🎉 All testing is looking good to me

MrMuzyk committed:
fix: various cr fixes

MrMuzyk committed:
fix: linter

amyevans reviewed (approved):
Looks like @mananjadhav's feedback was addressed and he already completed the checklist, and I'm 👍 on the code and testing, so I think we're all set to merge here!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9523224388) to staging by https://github.com/amyevans in version: 1.4.84-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9556355947) to production by https://github.com/AndrewGable in version: 1.4.84-3 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
